From d45dcbca9606bef4e96dfe6dba053af56d92e7e8 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Sun, 20 Mar 2022 12:49:58 +0200
Subject: [PATCH] [SW-57547] rom: add cold rst flops for validation

Update the cold reset flops for validation pass and fail. This
is required to skip re-validation of the image once validated.

Change-Id: I7143ddcd8ce51567d2c587a666ce4b81bb232b2f
---

diff --git a/app/arcboot_rom.c b/app/arcboot_rom.c
index 76173d8..885c3ba 100644
--- a/app/arcboot_rom.c
+++ b/app/arcboot_rom.c
@@ -197,7 +197,7 @@
 	return (!arcboot_rom_global.validation_pass);
 }
 
-void arcboot_rom_validation_fail(void)
+__weak void arcboot_rom_validation_fail(void)
 {
 	uint32_t reg;
 
diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 08b71b4..13f6d79 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -251,9 +251,26 @@
 	boot_desc_get_img_desc(HL_FW_BOOT_PTR_SPI, img, desc, upd_status);
 }
 
+void arcboot_rom_validation_fail(void)
+{
+	struct gaudi3_cold_rst_data cold_rst_data;
+
+	/* validation failed, set recovery flag and clear validation */
+	cold_rst_data.data = regmap_readl(mmCOLD_RST_FLOPS_2);
+	cold_rst_data.recovery_flag = 1;
+	cold_rst_data.validation_flag = 0;
+	regmap_writel(cold_rst_data.data, mmCOLD_RST_FLOPS_2);
+}
+
 static void arcboot_rom_validation_pass(void)
 {
-	/* TODO: add vaildation pass logic */
+	struct gaudi3_cold_rst_data cold_rst_data;
+
+	/* validation pass, set validation flag and clear recovery */
+	cold_rst_data.data = regmap_readl(mmCOLD_RST_FLOPS_2);
+	cold_rst_data.recovery_flag = 0;
+	cold_rst_data.validation_flag = 1;
+	regmap_writel(cold_rst_data.data, mmCOLD_RST_FLOPS_2);
 }
 
 #define PPBOOT_MAX_IMAGE_SIZE	(2 * MAX_DMA_SIZE_CRC)
